pytorch_lightning/trainer/training_loop.py

```diff
@@ -685,69 +692,74 @@ def run_training_batch(self, batch, batch_idx, dataloader_idx):
+            signal=0,
+            # todo: Properly aggregate grad_norm accros opt_idx and split_idx
+            grad_norm_dict=grad_norm_dict,
+            training_step_output_for_epoch_end=batch_outputs,
+        )
+        return output
 
-                    # -------------------
-                    # calculate loss (train step + train step end)
-                    # -------------------
+    def run_optimization(self, batch_idx, split_idx, split_batch, opt_idx=None, optimizer=None):
+        # toggle model params + set info to logger_connector
+        self.run_train_split_start(split_idx, split_batch, opt_idx, optimizer)
 
-                    # automatic_optimization=True: perform dpp sync only when performing optimizer_step
-                    # automatic_optimization=False: don't block synchronization here
-                    with self.block_ddp_sync_behaviour():
-                        result = self.training_step_and_backward(
-                            split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
-                        )
+        result = AttributeDict()
+        if self.should_accumulate():
+            # For gradient accumulation
 
-                # ------------------------------
-                # BACKWARD PASS
-                # ------------------------------
-                # gradient update with accumulated gradients
-                else:
-                    if self.trainer.lightning_module.automatic_optimization:
+            # -------------------
+            # calculate loss (train step + train step end)
+            # -------------------
 
-                        def train_step_and_backward_closure():
-                            nonlocal result
-                            result = self.training_step_and_backward(
-                                split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
-                            )
-                            return None if result is None else result.loss
+            # automatic_optimization=True: perform dpp sync only when performing optimizer_step
+            # automatic_optimization=False: don't block synchronization here
+            with self.block_ddp_sync_behaviour():
+                result = self.training_step_and_backward(
+                    split_batch, batch_idx, opt_idx, optimizer, self.trainer.hiddens
+                )
 
-                        # optimizer step
-                        self.optimizer_step(optimizer, opt_idx, batch_idx, train_step_and_backward_closure)
+        # ------------------------------
+        # BACKWARD PASS
+        # ------------------------------
+        # gradient update with accumulated gradients
+        else:
+            if self.trainer.lightning_module.automatic_optimization:
 
-                    else:
-                        result = self.training_step(split_batch, batch_idx, opt_idx, self.trainer.hiddens)
+                def train_step_and_backward_closure():
```

Comment:
Small nits. Mind integrate the changes from this PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/6212/files#diff-6b21474ed45079f01dfa45ee3b9d40d23efe693c9c005ce897e25384ef425349R692.

Let's replace the closure by a partial there. We need to update the Lightning Optimizer there too:
https://github.com/PyTorchLightning/pytorch-lightning/pull/6212/files#diff-d931a456f290914901d2d1456ab3c85e63b39669d00723d624b988213dccbfdeR213

Reply:
I integrated the make_closure with partial here in this PR as you requested. 
I don't understand why the partial call it is useful, because all arguments are supplied so the returned function is not really partial.